File: 94_Binary_Tree_Inorder_Traversal.cpp
//Definition for a binary tree node.
struct TreeNode
{
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution
{
public:
    void trace(TreeNode* pNode, vector<int> &ret)
    {
        if(pNode == nullptr)
            return;
        trace(pNode->left, ret);
        ret.push_back(pNode->val);
        trace(pNode->right, ret);

    }

    vector<int> inorderTraversal(TreeNode* root)
    {
        vector<int> ret;
        trace(root, ret);
        return ret;
    }
};

File: 94_Binary_Tree_Inorder_Traversal.java
class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;
    TreeNode() {}
    TreeNode(int val) { this.val = val; }
    TreeNode(int val, TreeNode left, TreeNode right) {
    this.val = val;
    this.left = left;
    this.right = right;
    }
}

class Solution {
    public void trace(TreeNode Node, List<Integer> ret)
    {
        if(Node == null)
        {
            return;
        }
        trace(Node.left, ret);
        ret.add(Node.val);
        trace(Node.right, ret);
    }

    public List<Integer> inorderTraversal(TreeNode root) {
        List<Integer> ret = new ArrayList<Integer>();
        trace(root, ret);
        return ret;
    }
}

